rust/src/sans_io/read.rs

```diff
@@ -0,0 +1,1014 @@
+    data_section_hasher: Option<crc32fast::Hasher>,
+    calculated_data_section_crc: Option<u32>,
+    decompressors: HashMap<String, Box<dyn Decompressor>>,
+    options: LinearReaderOptions,
+    at_eof: bool,
+    last_write: Option<usize>,
+    failed: bool,
+}
+
+impl LinearReader {
+    pub fn new() -> Self {
+        Self::new_with_options(LinearReaderOptions::default())
+    }
+
+    pub fn new_with_options(options: LinearReaderOptions) -> Self {
+        LinearReader {
+            currently_reading: if options.skip_start_magic {
+                Record
+            } else {
+                StartMagic
+            },
+            from: File,
+            uncompressed: RWBuf::default(),
+            compressed: RWBuf::default(),
+            data_section_hasher: if options.validate_data_section_crc {
+                None
+            } else {
+                Some(crc32fast::Hasher::new())
+            },
+            calculated_data_section_crc: None,
+            decompressors: HashMap::new(),
+            last_write: None,
+            at_eof: false,
+            options,
+            failed: false,
+        }
+    }
+
+    /// Constructs a linear reader that will iterate through all records in a chunk.
+    pub(crate) fn for_chunk(header: ChunkHeader) -> McapResult<Self> {
+        let mut result = Self::new_with_options(
+            LinearReaderOptions::default()
+                .with_skip_end_magic(true)
+                .with_skip_start_magic(true)
+                .with_validate_chunk_crcs(true),
+        );
+        result.currently_reading = Record;
+        result.from = Chunk(ChunkState {
+            decompressor: result.get_decompressor(&header.compression)?,
+            hasher: Some(crc32fast::Hasher::new()),
+            crc: header.uncompressed_crc,
+            compressed_remaining: header.compressed_size,
+            padding_after_compressed_data: 0,
+        });
+        Ok(result)
+    }
+
+    fn get_decompressor(&mut self, name: &str) -> McapResult<Option<Box<dyn Decompressor>>> {
+        if let Some(decompressor) = self.decompressors.remove(name) {
+            return Ok(Some(decompressor));
+        }
+        match name {
+            #[cfg(feature = "zstd")]
+            "zstd" => Ok(Some(Box::new(zstd::ZstdDecoder::new()))),
+            #[cfg(feature = "lz4")]
+            "lz4" => Ok(Some(Box::new(lz4::Lz4Decoder::new()?))),
+            "" => Ok(None),
+            _ => Err(McapError::UnsupportedCompression(name.into())),
+        }
+    }
+    fn return_decompressor(&mut self, mut decompressor: Box<dyn Decompressor>) -> McapResult<()> {
+        decompressor.reset()?;
+        self.decompressors
+            .insert(decompressor.name().into(), decompressor);
+        Ok(())
+    }
+
+    /// Yields the next action the caller should take to progress through the file.
+    pub fn next_action(&mut self) -> Option<McapResult<ReadAction>> {
+        self.next_action_inner().transpose()
```

Comment:
The transpose here is because when a function returns `Option<Result<...` , you can't use `?` syntax to return errors.  Having an inner method that returns `Result<` lets me use intuitive error return syntax while writing the big state machine logic, then i can transpose it to provide the more idiomatic `Option<Result<` iterator return value.

Reply:
👍 might be nice to have a brief comment about this